### DISCUSSION TITLE
Rumble/vibation for game controller? (+link to stream from Wizard Fu)

### DISCUSSION THREAD

**Discussion by rudiHammad** [axmolengine/axmol#1241] [Q&A] [answered]
Hi,
I search in the discussions of axmol using keyword like "rumble", "vibration", "gamepad"... but didn't find anything.
I asked long time ago in cocos2dx about it, but the answers where referring to the vibration of the mobil device.
My question is about how to create and set vibrations on the gamepad controller. I browser axmol classes but didn't
see anything. 

WizardFu did it years ago in is game Songbringer: here is one of his streams about it
[https://www.youtube.com/watch?v=VQWlFwbgd18](https://www.youtube.com/watch?v=VQWlFwbgd18)  (copy paste the link, because clicking on it gives a 404 error apparently. )
I tried to follow along how he did but got a bit lost since it is not a tutorial, it is a stream where is is doing his game. Maybe someone
with more experience can follow it better.

Anyway, thoughts on how to implement that in axmol?

thanks

**Comment by DelinWorks** [comment]
maybe use a gamepad interfacing library that works with xbox controllers like Xinput, I use it in my c++ project and it does need a wrapper.

**Comment by rudiHammad** [comment]
can you provide some more info please? Specific library, any cpp tests, etc...
thx

**Reply by DelinWorks** [reply]
The solution I provided is a last resort, Unfortunately GLFW (the library that manages windowing and interfaces with game controllers and is what used in axmol) doesn't support game pad haptics (rumble/vibration) yet [GLFW issue](https://github.com/glfw/glfw/issues/57) so I'm proposing the use of Xinput, Xinput is provided through msvc win32 so if you're looking to support linux then you can't, Read through this article it has examples on how to implement and use the library [WIN32 GAMEPAD XINPUT](https://learn.microsoft.com/en-us/windows/win32/api/xinput/) and here are some [examples](https://learn.microsoft.com/en-us/windows/win32/xinput/getting-started-with-xinput) for it, of course it is advised to use a wrapper since working with WIN32 API is frustrating. if you find any, but If I were you I wouldn't really bother with learning and implementing another library just for controller rumble, when time comes to release your game you can then branch out to better libraries like steam-works api which has full support for controller.

**Comment by rudiHammad** [comment]
great, you saved me sometime there with that info. Thanks.
Yes, my plan is to publish on steam so I'll look into it when the game is ready. 
I'll leave this question open a couple of days in case anyone else has any other suggestion and then I'll close it.

(By the way, quick of topic: are there any plans for ps5 ports? I read in this forum that there is no plan for the Switch because of nintendo policy, but haven't ready anything related to ps5. )

**Reply by DelinWorks** [reply, marked as answer]
all of these platforms Nintendo switch and Playstation require an NDA which in no way allows this engine or any other to build and publish on these platforms since it's open source, if we somehow do get this NDA to work then a lot of programming has to be done for platform support and other stuff like rendering and input apis.
